include/vara/Feature/Constraint.h

```diff
@@ -0,0 +1,379 @@
+#ifndef VARA_FEATURE_CONSTRAINT_H
+#define VARA_FEATURE_CONSTRAINT_H
+
+#include "vara/Feature/Feature.h"
+
+#include "llvm/ADT/StringRef.h"
+
+#include <cassert>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <variant>
+
+namespace vara::feature {
+
+//===----------------------------------------------------------------------===//
+//                               Constraint
+//===----------------------------------------------------------------------===//
+
+class ConstraintVisitor;
+
+/// \brief Tree like representation of constraints between features
+class Constraint {
+public:
+  enum class ConstraintKind { CK_BINARY, CK_UNARY, CK_PRIMARY };
+
+  Constraint(ConstraintKind Kind) : Kind(Kind) {}
+  Constraint(const Constraint &) = delete;
+  Constraint &operator=(const Constraint &) = delete;
+  virtual ~Constraint() = default;
+
+  [[nodiscard]] ConstraintKind getKind() const { return Kind; };
+
+  void setParent(Constraint *P) { this->Parent = P; }
+
+  [[nodiscard]] Constraint *getParent() { return Parent; }
+
+  [[nodiscard]] Constraint *getRoot() {
+    auto *R = this;
+    while (R->Parent) {
+      R = R->Parent;
+    }
+    return R;
+  }
+
+  [[nodiscard]] virtual std::string toString() const = 0;
+
+  [[nodiscard]] virtual std::string toHTML() const { return toString(); }
+
+  virtual void accept(ConstraintVisitor &V) = 0;
+
+private:
+  ConstraintKind Kind;
```

Comment:
```suggestion
  const ConstraintKind Kind;
```
So we can guarantee that the kind does not change during the Lifetime of a constraint.